packages/cactus-plugin-bungee-hermes/src/main/typescript/strategy/strategy-besu.ts

```diff
@@ -92,7 +107,8 @@ export class StrategyBesu implements ObtainLedgerStrategy {
 
   async getAllAssetsKey(
     networkDetails: BesuNetworkDetails,
-    api: BesuApi,
+    connector: PluginLedgerConnectorBesu | undefined,
```

Comment:
In which way? We can have one or the other and also both (no reason for that but it chooses the connector if both are present), so in a normal case one will be undefined (it checks if they are present, if none are it will throw an error).

So if you have a different idea on this please explain me.

Reply:
In here it checks if in InvokeContract function if either are present.

```
 if (connector) {
      const response = await connector.invokeContract(parameters);
      if (response.callOutput) {
        return response.callOutput as string[];
      } else {
        throw new Error(
          `${StrategyBesu.CLASS_NAME}#invokeContract: contract ${parameters.contractName} method ${parameters.methodName} output is falsy`,
        );
      }
    } else if (api) {
      const response = await api.invokeContractV1(parameters);
      if (response.status >= 200 && response.status < 300) {
        if (response.data.callOutput) {
          return response.data.callOutput as string[];
        } else {
          throw new Error(
            `${StrategyBesu.CLASS_NAME}#invokeContract: contract ${parameters.contractName} method ${parameters.methodName} output is falsy`,
          );
        }
      }
      throw new Error(
        `${StrategyBesu.CLASS_NAME}#invokeContract: BesuAPI error with status ${response.status}: ` +
          response.data,
      );
    }
    throw new Error(
      `${StrategyBesu.CLASS_NAME}#invokeContract: BesuAPI or Connector were not defined`,
    );
```

Reply:
@LordKubaya Here's an example of how to validate the contract invocation output in a robust way that avoids most issues:

```typescript
  async invokeContract(...): Promise<string[]> {
    const fn = `${StrategyBesu.CLASS_NAME}#invokeContract()`;
    Checks.truthy(this.connector, `${fn} this.connector`);
    const response = await connector.invokeContract(parameters);
    if (!response) {
      throw new Error(`${fn} invokeContract() response is falsy`);
    }
    if (!response.callOutput) {
      throw new Error(`${fn} invokeContract() response.callOutput is falsy`);
    }
    const { callOutput } = response;
    if (!Array.isArray(callOutput)) {
      throw new Error(`${fn} invokeContract() callOutput not an array`);
    }
    const allItemsAreStrings = callOutput.every((x) => typeof x === "string");
    if (!allItemsAreStrings) {
      throw new Error(`${fn} invokeContract() callOutput has non-string items`);
    }
    return response.callOutput as string[];
  }
```

Reply:
@LordKubaya @RafaelAPB Here's a more comprehensive example that actually compiles and passes the tests:

The branch:
https://github.com/petermetz/cacti/tree/refactor-bungee-error-handling-and-input-validation

The specific commit on the branch that I just made:
https://github.com/hyperledger/cacti/commit/ad4fafbbe69c0a8ad930a5f3616b8298b2d079c2

The commit message duplicated here just to make sure it's easy to find:

---

refactor(bungee): besu strategy error handling and input validation

Notice a few things:
1. The http-errors-enhanced-cjs library is being used to encode information
about whether an error is consdiered (by us) as a user error or a bug in our
code (e.g. it signals who do we think are at fault which is the important
question when handling errors.). If we think the input was valid an we should
hae succeeded with the function execution, then we throw an InternalServerError.
Otherwise if we think the input was not in the required format, we throw a
BadRequestError notifying the user (or developer who is calling our function)
that they need to fix their input that they provided us with.
2. The error handling code and the output validation code are flattened out
as much as possible so that it's easier to read and understand what the code
does in what scenario and as the execution progresses the possible bad situations
are slowly but surely eliminated and by the end we are confident that the output
was in the expected format and that we can return it as-is and we don't need to
have any type casting at all.
3. I also added the `pluginRegistry` parameter to the bungee plugin's options
because this is an object that the API server passes in to every plugin that
it instanties which is meant to facilitate the cross-plugin interaction, e.g.
you can use in each plugin the passed in pluginRegistry to look up other plugins
instances directly and invoke them without needing an API client object for doing
so (which also adds a huge overhead of network latency)
The relevant piece of code to be aware of to see why this is working is in the
API server class, as pasted below where you can see that the pluginRegistry gets
appended to all the plugin options regardless of which kind of plugin it is:
```typescript
  private async instantiatePlugin(
    pluginImport: PluginImport,
    registry: PluginRegistry,
  ): Promise<ICactusPlugin> {
    const fnTag = `${this.className}#instantiatePlugin`;
    const { logLevel } = this.options.config;
    const { packageName, options } = pluginImport;
    this.log.info(`Creating plugin from package: ${packageName}`, options);
    const pluginOptions = { ...options, logLevel, pluginRegistry: registry };
```

Signed-off-by: Peter Somogyvari <peter.somogyvari@accenture.com>